src/main/java/teammates/ui/webapi/output/SearchCourseData.java

```diff
@@ -0,0 +1,43 @@
+package teammates.ui.webapi.output;
+
+/**
+ * Represents a Course search result.
+ */
+public class SearchCourseData extends ApiOutput {
```

Comment:
Why not just use `CourseData`?

Reply:
That was what we used at the start, however, we realized that we were returning a lot of unused fields, which made joining on the front end very verbose. The same logic applies to the rest.

Reply:
Although my reason may not be strong enough, I think there are several things that does not "smell good" here:

- Code duplication (`SearchCourseData.java` and `CourseData.java` etc.) One may change one place and forget to update another.
- **Uniform** REST endpoint. (This might be the top reason). Returning unused fields is not a big deal. Consider this: When you use GitHub REST API, do you complain about the unused fields?

> we were returning a lot of unused fields, which made joining on the front end very verbose

You can leave the unused fields untouched?

Reply:
I think the part on code duplication is the reason why we chose to abstract it out. While doing this migration, the coupling between the our search response and existing classes caused a bit of problems. For example, we needed `googleId`, which was not in `StudentData`. However, adding `googleId` in would have caused more problems, because some of the endpoints which use `StudentData` might not need `googleId` at all, and hence we would have to go to all usages and set it to null. 
I think this grokking of all occurrences and modifying them is more likely to cause regression bugs, so we decided to decouple them as we felt that this minor code duplication is a lesser evil.
I do get where you are coming from though. We initially did it this way but created new response classes for all  search endpoints a week back due to some messiness in implementation.

Reply:
When designing the REST endpoint, `*Data` is acted as DTO (Data Transfer Object) between the RESTFul endpoint and the logic layer. There is a usually a one-to-one mapping between `*Data` and `*Attribute`. If you do a quick search, you will find the layer does a lot of abstraction to avoid "hack" in the frontend.

> the coupling between the our search response and existing classes caused a bit of problems.

I don't view it as a coupling. It is rather a natural thing to return `StudentData` as you are searching for students.

> For example, we needed googleId, which was not in `StudentData`

To me, it is more like a defect in the endpoint. `googleId` should be in the `StudentData` so please fix this. 

> is more likely to cause regression bugs

I think the only place you need to check is `GetStudentAction`.

> some messiness in implementation.

In the existing code?

Reply:
Yea, there was some messiness, but this argument makes sense.